File: core/src/main/java/ru/surfstudio/android/core/ui/base/message/DefaultMessageController.java
package ru.surfstudio.android.core.ui.base.message;

import android.content.res.TypedArray;
import android.graphics.Color;
import android.support.annotation.ColorRes;
import android.support.annotation.Nullable;
import android.support.annotation.StringRes;
import android.support.design.widget.CoordinatorLayout;
import android.support.design.widget.Snackbar;
import android.support.v7.app.AppCompatActivity;
import android.view.Gravity;
import android.view.View;
import android.widget.Toast;

import ru.surfstudio.android.core.R;
import ru.surfstudio.android.core.ui.base.dagger.provider.ActivityProvider;
import ru.surfstudio.android.core.ui.base.dagger.provider.FragmentProvider;
import ru.surfstudio.android.dagger.scope.PerScreen;

/**
 * Для нахождения view использует fragment и затем activity провайдеры
 */
@PerScreen
public class DefaultMessageController implements MessageController {

    private static final int ILLEGAL_COLOR = Color.TRANSPARENT;
    private static final int DEFAULT_TOAST_GRAVITY = Gravity.BOTTOM;

    private final ActivityProvider activityProvider;

    @Nullable
    private final FragmentProvider fragmentProvider;

    @Nullable
    private Integer snackBarBackgroundColor;
    private Toast toast;

    public DefaultMessageController(ActivityProvider activityProvider,
                                    @Nullable FragmentProvider fragmentProvider) {
        this.activityProvider = activityProvider;
        this.fragmentProvider = fragmentProvider;

        TypedArray typedArray = activityProvider.get()
                .obtainStyledAttributes(new int[]{R.attr.snackBarBackgroundColor});
        try {
            int color = typedArray.getColor(0, ILLEGAL_COLOR);
            if (color != ILLEGAL_COLOR) {
                snackBarBackgroundColor = color;
            }
        } catch (UnsupportedOperationException ignored) {
            // ignored
        }

        typedArray.recycle();
    }

    public DefaultMessageController(ActivityProvider activityProvider) {
        this(activityProvider, null);
    }

    @Override
    public void show(@StringRes int stringId) {
        show(getView(), stringId);
    }

    public void show(View view, int stringId) {
        show(view, view.getResources().getString(stringId));
    }

    @Override
    public void show(String message) {
        show(getView(), message);
    }

    @Override
    public void showToast(@StringRes int stringId) {
        showToast(getView().getResources().getString(stringId), DEFAULT_TOAST_GRAVITY);
    }

    @Override
    public void showToast(@StringRes int stringId, int gravity) {
        showToast(getView().getResources().getString(stringId), gravity);
    }

    @Override
    public void showToast(String message) {
        showToast(message, DEFAULT_TOAST_GRAVITY);
    }

    @Override
    public void showToast(String message, int gravity) {
        if (toast != null) {
            toast.cancel();
        }
        toast = Toast.makeText(getView().getContext(), message, Toast.LENGTH_LONG);
        toast.setGravity(gravity, 0, 0);
        toast.show();
    }

    public void show(View view, String message) {
        Snackbar sb = Snackbar.make(view, message, Snackbar.LENGTH_LONG);
        if (snackBarBackgroundColor != null) {
            sb.getView().setBackgroundColor(snackBarBackgroundColor);
        }

        sb.show();
    }

    public void show(String message, int backgroundColor) {
        show(getView(), message, backgroundColor);
    }

    public void show(View view, String message, int backgroundColor) {
        Snackbar sb = Snackbar.make(view, message, Snackbar.LENGTH_LONG);
        sb.getView().setBackgroundColor(backgroundColor);
        sb.show();
    }

    public void show(@StringRes int stringId,
                     @ColorRes int backgroundColor,
                     @StringRes int actionStringId,
                     @ColorRes int buttonColor) {
        View view = getView();
        Snackbar sb = Snackbar.make(view, view.getResources().getString(stringId), Snackbar.LENGTH_INDEFINITE);
        sb.getView().setBackgroundColor(view.getResources().getColor((backgroundColor)));
        sb.setAction(actionStringId, v -> sb.dismiss());
        sb.setActionTextColor(view.getResources().getColor(buttonColor));
        sb.show();
    }

    public void show(@StringRes int stringId,
                     @StringRes int actionStringId,
                     @ColorRes int buttonColor,
                     View.OnClickListener listener) {
        View view = getView();
        Snackbar sb = Snackbar.make(view, view.getResources().getString(stringId), Snackbar.LENGTH_INDEFINITE);
        if (snackBarBackgroundColor != null) {
            sb.getView().setBackgroundColor(snackBarBackgroundColor);
        }
        sb.setAction(actionStringId, listener);
        sb.setActionTextColor(view.getResources().getColor(buttonColor));
        sb.show();
    }

    /**
     * Порядок поиска подходящей корневой вью для SnackBar происходит со следующим приоритетом:
     * R.id.snackbar_container во фрагменте, должен быть FrameLayout
     * R.id.coordinator во фрагменте, должен быть CoordinatorLayout
     * R.id.snackbar_container в активитиб должен быть CoordinatorLayout или FrameLayout
     * R.id.coordinator в активити, должен быть CoordinatorLayout
     * android.R.id.content активити
     * <p>
     * Для того, чтобы срабатывал Behavior на появление Snackbar,
     * нужно чтобы найденая View была {@link CoordinatorLayout}
     */
    protected View getView() {
        View v = getViewFromFragment(fragmentProvider);
        if (v == null) {
            v = getViewFromActivity(activityProvider);
        }

        return v;
    }

    private View getViewFromFragment(FragmentProvider fragmentProvider) {
        if (fragmentProvider == null) {
            return null;
        }
        View fragmentView = fragmentProvider.get().getView();
        if (fragmentView == null) {
            return null;
        }
        View v = fragmentView.findViewById(R.id.snackbar_container);
        if (v == null) {
            v = fragmentView.findViewById(R.id.coordinator);
        }
        return v;
    }

    private View getViewFromActivity(ActivityProvider activityProvider) {
        AppCompatActivity activity = activityProvider.get();
        View v = activity.findViewById(R.id.snackbar_container);
        if (v == null) {
            v = activity.findViewById(R.id.coordinator);
        }
        if (v == null) {
            v = activity.findViewById(android.R.id.content);
        }
        return v;
    }
}

File: message-controller/lib-message-controller/src/main/java/ru/surfstudio/android/message/DefaultMessageController.kt
/*
  Copyright (c) 2018-present, SurfStudio LLC.

  Licensed under the Apache License, Version 2.0 (the "License");
  you may not use this file except in compliance with the License.
  You may obtain a copy of the License at

  http://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.
 */
package ru.surfstudio.android.message

import android.annotation.SuppressLint
import android.graphics.Color
import android.view.View
import android.widget.Toast
import androidx.annotation.ColorInt
import androidx.annotation.ColorRes
import androidx.annotation.StringRes
import androidx.appcompat.app.AppCompatActivity
import androidx.coordinatorlayout.widget.CoordinatorLayout
import androidx.core.content.ContextCompat
import com.google.android.material.snackbar.Snackbar
import ru.surfstudio.android.core.ui.provider.FragmentProvider
import ru.surfstudio.android.core.ui.provider.Provider

/**
 * Базовый класс контроллера отображения сообщений
 * Для нахождения view использует fragment и затем activity провайдеры
 */
class DefaultMessageController @JvmOverloads constructor(
        val activityProvider: Provider<AppCompatActivity>,
        val fragmentProvider: FragmentProvider? = null
) : MessageController {

    @ColorInt
    private var snackBarBackgroundColor: Int? = null
    @ColorInt
    private var snackBarActionButtonColor: Int? = null
    private var toast: Toast? = null
    private var snackbar: Snackbar? = null

    init {
        val typedArray = activityProvider.get()
                .obtainStyledAttributes(intArrayOf(R.attr.snackBarBackgroundColor, R.attr.snackBarActionButtonColor))
        try {
            if (typedArray.hasValue(R.styleable.Core_snackBarBackgroundColor)) {
                snackBarBackgroundColor = typedArray.getColor(R.styleable.Core_snackBarBackgroundColor, Color.TRANSPARENT)
            }
            if (typedArray.hasValue(R.styleable.Core_snackBarActionButtonColor)) {
                snackBarActionButtonColor = typedArray.getColor(R.styleable.Core_snackBarActionButtonColor, Color.TRANSPARENT)
            }
        } catch (ignored: UnsupportedOperationException) {
            // ignored
        } finally {
            typedArray.recycle()
        }
    }

    override fun show(
            message: CharSequence,
            @ColorRes
            backgroundColorResId: Int?,
            @StringRes
            actionResId: Int?,
            @ColorRes
            actionColorResId: Int?,
            duration: Int,
            listener: (view: View) -> Unit
    ) {
        show(
                SnackParams(
                        message = message,
                        backgroundColorResId = backgroundColorResId ?: 0,
                        actionResId = actionResId ?: 0,
                        actionColorResId = actionColorResId ?: 0,
                        duration = duration
                ),
                listener
        )
    }

    override fun show(
            @StringRes
            messageResId: Int,
            @ColorRes
            backgroundColorResId: Int?,
            @StringRes
            actionResId: Int?,
            @ColorRes
            actionColorResId: Int?,
            duration: Int,
            listener: (view: View) -> Unit
    ) {
        show(
                SnackParams(
                        messageResId = messageResId,
                        backgroundColorResId = backgroundColorResId ?: 0,
                        actionResId = actionResId ?: 0,
                        actionColorResId = actionColorResId ?: 0,
                        duration = duration
                ),
                listener
        )
    }

    override fun show(params: SnackParams, actionListener: (view: View) -> Unit) {
        val activity = activityProvider.get()
        val message = if (params.messageResId != 0) {
            activity.getString(params.messageResId)
        } else {
            params.message
        }
        var duration = params.duration
        if (duration != Snackbar.LENGTH_SHORT
                && duration != Snackbar.LENGTH_LONG
                && duration != Snackbar.LENGTH_INDEFINITE) {
            duration = Snackbar.LENGTH_SHORT
        }
        snackbar = Snackbar.make(getView(), message, duration).apply {
            var backgroundColor: Int? = if (params.backgroundColorResId != 0) {
                ContextCompat.getColor(activity, params.backgroundColorResId)
            } else {
                params.backgroundColor
            }
            if (backgroundColor == null) {
                backgroundColor = snackBarBackgroundColor
            }
            if (backgroundColor != null) {
                view.setBackgroundColor(backgroundColor)
            }
            val actionText = if (params.actionResId != 0) {
                activity.getString(params.actionResId)
            } else {
                params.action
            }
            if (actionText.isNotEmpty()) {
                setAction(actionText) { view -> actionListener.invoke(view) }
            }
            var actionButtonColor = if (params.actionColorResId != 0) {
                ContextCompat.getColor(activity, params.actionColorResId)
            } else {
                params.actionColor
            }
            if (actionButtonColor == null) {
                actionButtonColor = snackBarActionButtonColor
            }
            if (actionButtonColor != null) {
                setActionTextColor(actionButtonColor)
            }
            show()
        }
    }

    override fun closeSnack() {
        snackbar?.dismiss()
    }

    override fun showToast(
            @StringRes messageResId: Int,
            gravity: Int?,duration: Int
    ) {
        showToast(
                ToastParams(messageResId = messageResId ?: 0,
                        gravity = gravity,
                        duration = duration)
        )
    }

    override fun showToast(
            message: CharSequence,
            gravity: Int?,
            duration: Int
    ) {
        showToast(
                ToastParams(message = message,
                        gravity = gravity,
                        duration = duration)
        )
    }

    @SuppressLint("ShowToast")
    override fun showToast(params: ToastParams) {
        toast?.cancel()
        val activity = activityProvider.get()
        val toast: Toast
        var duration = params.duration
        if (duration != Toast.LENGTH_SHORT && duration != Toast.LENGTH_LONG) {
            duration = Toast.LENGTH_SHORT
        }
        if (params.customView == null) {
            val message = if (params.messageResId != 0) {
                activity.getString(params.messageResId)
            } else {
                params.message
            }
            toast = Toast.makeText(activity, message, duration)
        } else {
            toast = Toast(activity)
            toast.view = params.customView
            toast.duration = params.duration
        }
        params.gravity?.let {
            toast.setGravity(params.gravity, params.xOffset, params.yOffset)
        }
        toast.show()
    }

    /**
     * Порядок поиска подходящей корневой вью для SnackBar происходит со следующим приоритетом:
     * R.id.snackbar_container во фрагменте, должен быть FrameLayout
     * R.id.coordinator во фрагменте, должен быть CoordinatorLayout
     * R.id.snackbar_container в активити, должен быть CoordinatorLayout или FrameLayout
     * R.id.coordinator в активити, должен быть CoordinatorLayout
     * android.R.id.content активити
     *
     *
     * Для того, чтобы срабатывал Behavior на появление Snackbar,
     * нужно чтобы найденая View была [CoordinatorLayout]
     */
    protected fun getView(): View {
        return getViewFromFragment(fragmentProvider)
                ?: getViewFromActivity(activityProvider)
                ?: throw ViewForSnackbarNotFoundException("ViewForSnackbarNotFound")
    }

    private fun getViewFromFragment(fragmentProvider: FragmentProvider?): View? {
        val fragmentView = fragmentProvider?.get()?.view ?: return null
        return fragmentView.findViewById(R.id.snackbar_container)
                ?: fragmentView.findViewById(R.id.coordinator)
    }

    private fun getViewFromActivity(activityProvider: Provider<AppCompatActivity>): View? {
        val activity = activityProvider.get()
        return activity.findViewById(R.id.snackbar_container)
                ?: activity.findViewById(R.id.coordinator)
                ?: activity.findViewById(android.R.id.content)
    }
}